homeassistant/components/zwave_mqtt/binary_sensor.py

```diff
@@ -0,0 +1,400 @@
+        # Index 14: Siren
+        NOTIFICATION_TYPE: ValueIndex.NOTIFICATION_SIREN,
+        NOTIFICATION_VALUES: [1],
+        NOTIFICATION_DEVICE_CLASS: DEVICE_CLASS_SOUND,
+    },
+    {
+        # Index 15: Water valve
+        # ignore non-boolean values
+        NOTIFICATION_TYPE: ValueIndex.NOTIFICATION_WATER_VALVE,
+        NOTIFICATION_VALUES: [3, 4],
+        NOTIFICATION_DEVICE_CLASS: DEVICE_CLASS_PROBLEM,
+    },
+    {
+        # Index 16: Weather
+        NOTIFICATION_TYPE: ValueIndex.NOTIFICATION_WEATHER,
+        NOTIFICATION_VALUES: [1, 2],
+        NOTIFICATION_DEVICE_CLASS: DEVICE_CLASS_PROBLEM,
+    },
+    {
+        # Index 17: Irrigation
+        # ignore non-boolean values
+        NOTIFICATION_TYPE: ValueIndex.NOTIFICATION_IRRIGATION,
+        NOTIFICATION_VALUES: [1, 2, 3, 4, 5],
+        NOTIFICATION_DEVICE_CLASS: None,
+    },
+    {
+        # Index 18: Gas
+        NOTIFICATION_TYPE: ValueIndex.NOTIFICATION_GAS,
+        NOTIFICATION_VALUES: [1, 2, 3, 4],
+        NOTIFICATION_DEVICE_CLASS: DEVICE_CLASS_GAS,
+    },
+    {
+        # Index 18: Gas
+        NOTIFICATION_TYPE: ValueIndex.NOTIFICATION_GAS,
+        NOTIFICATION_VALUES: [6],
+        NOTIFICATION_DEVICE_CLASS: DEVICE_CLASS_PROBLEM,
+    },
+]
+
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up Z-Wave binary_sensor from config entry."""
+
+    @callback
+    def async_add_binary_sensor(values):
+        """Add Z-Wave Binary Sensor."""
+        sensors_to_add = []
+
+        if values.primary.type == ValueType.LIST:
+            # we convert Notification values into binary sensors
+            # https://github.com/OpenZWave/open-zwave/blob/master/config/NotificationCCTypes.xml
+            for list_value in values.primary.value["List"]:
+                # check if we have a mapping for this value
+                for item in NOTIFICATION_SENSORS:
+                    if item[NOTIFICATION_TYPE] != values.primary.index:
+                        continue
+                    if list_value["Value"] not in item[NOTIFICATION_VALUES]:
+                        continue
+                    sensors_to_add.append(
+                        ZWaveListValueSensor(
+                            # required values
+                            values,
+                            list_value["Value"],
+                            item[NOTIFICATION_DEVICE_CLASS],
+                            # optional values
+                            item.get(NOTIFICATION_SENSOR_ENABLED, True),
+                            item.get(NOTIFICATION_OFF_VALUE, NOTIFICATION_VALUE_CLEAR),
+                        )
+                    )
+
+        elif values.primary.type == ValueType.BOOL:
+            # classic/legacy binary sensor
+            sensors_to_add.append(ZWaveBinarySensor(values))
+        else:
+            # should not happen but log it just in case
+            _LOGGER.warning("Sensor not implemented for value %s", values.primary.label)
```

Comment:
This can only happen if we write the discovery schema incorrectly in the future. Either we should remove this case or raise an exception I think. If we keep it we should test it by patching the discovery schema.

Reply:
Instead of the if else block we can make a dictionary that maps the discovery type to the binary_sensor entity class. We look up the dict with the primary value type.